lib/redux/actions/async.js

```diff
@@ -422,105 +422,134 @@ export function doVersionCheck() {
-    );
-
-    if (!_.isEmpty(targetTimezones[uploadTargetUser]) &&
-      !_.isEmpty(targetDevices[uploadTargetUser])) {
-      dispatch(syncActions.setPage(pages.MAIN));
     }
-
-    const uploadsByUser = actionUtils.getDeviceTargetsByUser(usersWithTargets);
-
-    dispatch(syncActions.setUploads(uploadsByUser));
   };
 }
 
 export function retrieveTargetsFromStorage() {
   return (dispatch, getState) => {
     const { devices, uploadTargetUser } = getState();
     const { api, localStore } = services;
+    let fromLocalStore = false;
+
     dispatch(syncActions.retrieveUsersTargetsFromStorage());
-    const targets = localStore.getItem('devices');
-    if (targets === null) {
-      return dispatch(syncActions.setPage(pages.SETTINGS));
-    }
-    else {
+    let targets = localStore.getItem('devices');
+    if (targets !== null) {
+      fromLocalStore = true;
+      // wipe out the deprecated 'devices' localStore key
+      localStore.removeItem('devices');
       const uploadsByUser = actionUtils.getDeviceTargetsByUser(targets);
-
       dispatch(syncActions.setUploads(uploadsByUser));
+      dispatch(syncActions.setUsersTargets(targets));
     }
-    dispatch(syncActions.setUsersTargets(targets));
 
-    if (targets[uploadTargetUser] != null) {
-      const userTargets = targets[uploadTargetUser];
-      const targetDeviceKeys = _.pluck(userTargets, 'key');
+    const { targetDevices, targetTimezones } = getState();
+
+    // redirect based on having a supported device
+    if (!_.isEmpty(_.get(targetDevices, uploadTargetUser))) {
+      let usersWithTargets = {};
+      _.forOwn(targetDevices, (devicesArray, userId) => {
+        usersWithTargets[userId] = _.map(devicesArray, (deviceKey) => {
+          return {key: deviceKey};
+        });
+      });
+      _.forOwn(targetTimezones, (timezoneName, userId) => {
+        usersWithTargets[userId] = _.map(usersWithTargets[userId], (target) => {
+          if (timezoneName != null) {
+            target.timezone = timezoneName;
+          }
+          return target;
+        });
+      });
+      const targetDeviceKeys = targetDevices[uploadTargetUser];
       const supportedDeviceKeys = Object.keys(devices);
       const atLeastOneDeviceSupportedOnSystem = _.some(targetDeviceKeys, (key) => {
         return _.includes(supportedDeviceKeys, key);
       });
-      let timezones = [];
-      _.each(userTargets, (target) => {
-        if (target.timezone) {
-          timezones.push(target.timezone);
+
+      if(!fromLocalStore){
+        const uploadsByUser = actionUtils.getDeviceTargetsByUser(usersWithTargets);
+        dispatch(syncActions.setUploads(uploadsByUser));
+      } else {
+        if (!_.isEmpty(targetDevices[uploadTargetUser]) || !_.isEmpty(targetTimezones[uploadTargetUser])) {
+          dispatch(syncActions.updateProfileRequest());
+          api.user.updateProfile(uploadTargetUser, {
+            'patient': {
+              'targetDevices': targetDevices[uploadTargetUser],
```

Comment:
What if there aren't `targetDevices` or `targetTimezones` for the `uploadTargetUser`. Are you going to cause an API error here by pushing up an `undefined`? We don't expect one of these to ever be missing, so one strategy could be to change the `if` condition to require both.